src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -0,0 +1,609 @@
+            const interpreter =
+                option && isInterpreter(option)
+                    ? option
+                    : resource && this.extensionChecker.isPythonExtensionInstalled
+                    ? await this.interpreterService.getActiveInterpreter(resource)
+                    : undefined;
+
+            // Find the preferred kernel index from the list.
+            const notebookMetadata = option && !isInterpreter(option) ? option : undefined;
+            const preferred = findPreferredKernelIndex(kernels, resource, [], notebookMetadata, interpreter, undefined);
+            if (preferred >= 0) {
+                return kernels[preferred];
+            }
+        } catch (e) {
+            traceError(`findKernel crashed: ${e} ${e.stack}`);
+            return undefined;
+        }
+    }
+
+    // Search all our local file system locations for installed kernel specs and return them
+    @captureTelemetry(Telemetry.KernelListingPerf)
+    public async listKernels(
+        resource: Resource,
+        cancelToken?: CancellationToken
+    ): Promise<LocalKernelConnectionMetadata[]> {
+        try {
+            // Get an id for the workspace folder, if we don't have one, use the fsPath of the resource
+            const workspaceFolderId = this.workspaceService.getWorkspaceFolderIdentifier(
+                resource,
+                resource?.fsPath || this.workspaceService.rootPath
+            );
+
+            // If we have not already searched for this resource, then generate the search
+            if (workspaceFolderId && !this.workspaceToMetadata.has(workspaceFolderId)) {
+                this.workspaceToMetadata.set(workspaceFolderId, this.findResourceKernelMetadata(resource, cancelToken));
+            }
+
+            this.writeCache().ignoreErrors();
+
+            // ! as the has and set above verify that we have a return here
+            const items = await this.workspaceToMetadata.get(workspaceFolderId)!;
+            traceInfoIf(
+                !!process.env.VSC_JUPYTER_LOG_KERNEL_OUTPUT,
+                `Kernel specs for ${resource?.toString() || 'undefined'} are \n ${JSON.stringify(items)}`
+            );
+            return items;
+        } catch (e) {
+            traceError(`List kernels failed: ${e} ${e.stack}`);
+            throw e;
+        }
+    }
+
+    // This should return a WRITABLE place that jupyter will look for a kernel as documented
+    // here: https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs
+    public async getKernelSpecRootPath(): Promise<string | undefined> {
+        if (this.platformService.isWindows) {
+            return tryGetRealPath(path.join(this.pathUtils.home, winJupyterPath));
+        } else if (this.platformService.isMac) {
+            return tryGetRealPath(path.join(this.pathUtils.home, macJupyterPath));
+        } else {
+            return tryGetRealPath(path.join(this.pathUtils.home, linuxJupyterPath));
```

Comment:
No need for `getRealPath` for any of these.

Reply:
Yeah probably not for the mac/linux paths.